In [41]:
import pandas as pd
import collections

import matplotlib
import matplotlib.pyplot as plt
import numpy as np

%matplotlib inline

from tqdm import tqdm

In [13]:
large_tr_news = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDlarge_train/integrated_news.csv')
large_dev_news = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDlarge_dev/integrated_news.csv')
demo_tr_news = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDdemo_train/integrated_news.csv')
demo_dev_news = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDdemo_dev/integrated_news.csv')

In [98]:
print('MINDlarge_train 수집된 데이터 :', large_tr_news.shape)
print('MINDlarge_dev 수집된 데이터 :', large_dev_news.shape)
print('MINDdemo_train 수집된 데이터 :', demo_tr_news.shape)
print('MINDdemo_train 수집된 데이터 :', demo_dev_news.shape)

MINDlarge_train 수집된 데이터 : (101527, 12)
MINDlarge_dev 수집된 데이터 : (72023, 12)
MINDdemo_train 수집된 데이터 : (26740, 12)
MINDdemo_train 수집된 데이터 : (18723, 12)


In [15]:
large_tr_behavior = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDlarge_train/behaviors.tsv', delimiter='\t', header=None)
large_dev_behavior = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDlarge_dev/behaviors.tsv', delimiter='\t', header=None)
# demo_tr_behavior = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDdemo_train/behaviors.tsv', delimiter='\t', header=None)
# demo_dev_behavior = pd.read_csv('/Users/baeyuna/Documents/SNU_DLab/Data/MINDdemo_dev/behaviors.tsv', delimiter='\t', header=None)

# 1. large_tr

In [16]:
large_tr_news.head()

5 Title Date Image Content  \
0                                                 []   NaN  NaN   NaN     NaN   
1                                                 []   NaN  NaN   NaN     NaN   
2  [{"Label": "Freddie Kitchens", "Type": "P", "W...   NaN  NaN   NaN     NaN   
3  [{"Label": "Forest Park\u2013DeBaliviere stati...   NaN  NaN   NaN     NaN   
4  [{"Label": "Animal control service", "Type": "...   NaN  NaN   NaN     NaN   

         0          1                                2  \
0  N110434  lifestyle  lifestylefamilyandrelationships   
1   N74207    finance                    finance-video   
2  N113363     sports                     football_nfl   
3  N118418       news                           newsus   
4  N102010       news                           newsus   

                                                   3  \
0           The 50 Most Common Last Names in America   
1  The Price You Pay: The spiraling cost of colle...   
2  Baker Mayfield injury update: 'No doubt' Brown...   
3  The security has been stepped up': Forest Park...   
4  [We] walk dogs and pick up poo for free | Why ...   

                                                   4  \
0  What's in a name?\thttps://assets.msn.com/labs...   
1      https://assets.msn.com/labs/mind/AAIKUGl.html   
2  He'll be in better shape physically and he'll ...   
3      https://assets.msn.com/labs/mind/AAJncPt.html   
4      https://assets.msn.com/labs/mind/AAJUQ4w.html   

                                                   6    7  
0                                                NaN  NaN  
1                                                 []  NaN  
2                                                NaN  NaN  
3  [{"Label": "St. Louis", "Type": "G", "Wikidata...  NaN  
4  [{"Label": "Animal control service", "Type": "...  NaN

In [17]:
print(large_tr_news.shape)
complete_large_tr_news = large_tr_news.loc[~large_tr_news['Title'].isnull()]
print(complete_large_tr_news.shape)

print('주소 이전한 페이지 수 :', large_tr_news.shape[0] - complete_large_tr_news.shape[0])
complete_large_tr_news.head()

(101527, 12)
(101521, 12)
주소 이전한 페이지 수 : 6


5  \
6   https://assets.msn.com/labs/mind/AABmf2I.html   
7   https://assets.msn.com/labs/mind/AAISxPN.html   
8   https://assets.msn.com/labs/mind/AAEKnO1.html   
9   https://assets.msn.com/labs/mind/AADlomf.html   
10  https://assets.msn.com/labs/mind/AAGBR44.html   

                                                Title                  Date  \
6     Walmart Slashes Prices on Last-Generation iPads  2019-10-28T17:38:04Z   
7   Dispose of unwanted prescription drugs on Satu...  2019-10-25T17:31:42Z   
8   Check Houston traffic map for current road con...  2019-11-26T12:55:03Z   
9   How to record your screen on Windows, macOS, i...  2019-11-06T16:06:25Z   
10  This Guy Altered His Diet and Training to Drop...  2019-08-31T11:37:11Z   

                                                Image  \
6   https://img-s-msn-com.akamaized.net/tenant/amp...   
7   https://img-s-msn-com.akamaized.net/tenant/amp...   
8   https://img-s-msn-com.akamaized.net/tenant/amp...   
9   https://img-s-msn-com.akamaized.net/tenant/amp...   
10  https://img-s-msn-com.akamaized.net/tenant/amp...   

                                              Content       0       1  \
6   This year, Walmart’s not waiting until Black...  N45436    news   
7   CINCINNATI (FOX19) - If you have expired, unus...  N86255  health   
8   Get real-time traffic data from Houston's road...   N5771  health   
9   There are lots of reasons to record your scree...  N51947    news   
10  Brandon Reid always felt like “the fat kid.â...  N17957  health   

                           2  \
6   newsscienceandtechnology   
7                    medical   
8                     cardio   
9   newsscienceandtechnology   
10                weightloss   

                                                    3  \
6     Walmart Slashes Prices on Last-Generation iPads   
7   Dispose of unwanted prescription drugs during ...   
8   Check Houston traffic map for current road con...   
9   How to record your screen on Windows, macOS, i...   
10  This Guy Altered His Diet and Training to Drop...   

                                                    4  \
6   Apple's new iPad releases bring big deals on l...   
7                                                 NaN   
8                                                 NaN   
9   The easiest way to record what's happening on ...   
10  Take Brandon Reid's advice: "Don't worry what ...   

                                                    6  \
6   [{"Label": "IPad", "Type": "J", "WikidataId": ...   
7   [{"Label": "Drug Enforcement Administration", ...   
8   [{"Label": "Houston", "Type": "G", "WikidataId...   
9   [{"Label": "Microsoft Windows", "Type": "J", "...   
10                                                 []   

                                                    7  
6   [{"Label": "IPad", "Type": "J", "WikidataId": ...  
7                                                  []  
8                                                  []  
9                                                  []  
10                                                 []

In [18]:
no_image_large_tr = complete_large_tr_news.loc[complete_large_tr_news['Image'].isnull()]
image_large_tr = complete_large_tr_news.loc[~complete_large_tr_news['Image'].isnull()]
print('이미지가 없는 기사 수 (large_tr):', no_image_large_tr.shape[0])
print('이미지가 있는 기사 수 (large_tr):', image_large_tr.shape[0])
print() 
print('기사 속 이미지 포함 비율 :', (image_large_tr.shape[0]/complete_large_tr_news.shape[0])*100)

이미지가 없는 기사 수 (large_tr): 20307
이미지가 있는 기사 수 (large_tr): 81214

기사 속 이미지 포함 비율 : 79.99724194994138


In [19]:
large_tr_behavior.head() 

0        1                       2  \
0  1   U87243  11/10/2019 11:30:54 AM   
1  2  U598644   11/12/2019 1:45:29 PM   
2  3  U532401  11/13/2019 11:23:03 AM   
3  4  U593596  11/12/2019 12:24:09 PM   
4  5  U239687   11/14/2019 8:03:01 PM   

                                                   3  \
0  N8668 N39081 N65259 N79529 N73408 N43615 N2937...   
1  N56056 N8726 N70353 N67998 N83823 N111108 N107...   
2  N128643 N87446 N122948 N9375 N82348 N129412 N5...   
3  N31043 N39592 N4104 N8223 N114581 N92747 N1207...   
4  N65250 N122359 N71723 N53796 N41663 N41484 N11...   

                                                   4  
0  N78206-0 N26368-0 N7578-0 N58592-0 N19858-0 N5...  
1  N47996-0 N82719-0 N117066-0 N8491-0 N123784-0 ...  
2              N103852-0 N53474-0 N127836-0 N47925-1  
3  N38902-0 N76434-0 N71593-0 N100073-0 N108736-0...  
4  N76209-0 N48841-0 N67937-0 N62235-0 N6307-0 N3...

In [26]:
def impression_lst(x):
    return x.split(' ') 

In [25]:
large_tr_behavior.columns = ['Impression ID', 'User ID', 'Impression Time', 'Uesr Click History', 'Impression News']
large_tr_behavior.head() 

Impression ID  User ID         Impression Time  \
0              1   U87243  11/10/2019 11:30:54 AM   
1              2  U598644   11/12/2019 1:45:29 PM   
2              3  U532401  11/13/2019 11:23:03 AM   
3              4  U593596  11/12/2019 12:24:09 PM   
4              5  U239687   11/14/2019 8:03:01 PM   

                                  Uesr Click History  \
0  N8668 N39081 N65259 N79529 N73408 N43615 N2937...   
1  N56056 N8726 N70353 N67998 N83823 N111108 N107...   
2  N128643 N87446 N122948 N9375 N82348 N129412 N5...   
3  N31043 N39592 N4104 N8223 N114581 N92747 N1207...   
4  N65250 N122359 N71723 N53796 N41663 N41484 N11...   

                                     Impression News  
0  N78206-0 N26368-0 N7578-0 N58592-0 N19858-0 N5...  
1  N47996-0 N82719-0 N117066-0 N8491-0 N123784-0 ...  
2              N103852-0 N53474-0 N127836-0 N47925-1  
3  N38902-0 N76434-0 N71593-0 N100073-0 N108736-0...  
4  N76209-0 N48841-0 N67937-0 N62235-0 N6307-0 N3...

In [27]:
large_tr_behavior['Impression_News_lst'] = large_tr_behavior['Impression News'].apply(impression_lst)
large_tr_behavior.head()  

Impression ID  User ID         Impression Time  \
0              1   U87243  11/10/2019 11:30:54 AM   
1              2  U598644   11/12/2019 1:45:29 PM   
2              3  U532401  11/13/2019 11:23:03 AM   
3              4  U593596  11/12/2019 12:24:09 PM   
4              5  U239687   11/14/2019 8:03:01 PM   

                                  Uesr Click History  \
0  N8668 N39081 N65259 N79529 N73408 N43615 N2937...   
1  N56056 N8726 N70353 N67998 N83823 N111108 N107...   
2  N128643 N87446 N122948 N9375 N82348 N129412 N5...   
3  N31043 N39592 N4104 N8223 N114581 N92747 N1207...   
4  N65250 N122359 N71723 N53796 N41663 N41484 N11...   

                                     Impression News  \
0  N78206-0 N26368-0 N7578-0 N58592-0 N19858-0 N5...   
1  N47996-0 N82719-0 N117066-0 N8491-0 N123784-0 ...   
2              N103852-0 N53474-0 N127836-0 N47925-1   
3  N38902-0 N76434-0 N71593-0 N100073-0 N108736-0...   
4  N76209-0 N48841-0 N67937-0 N62235-0 N6307-0 N3...   

                                 Impression_News_lst  
0  [N78206-0, N26368-0, N7578-0, N58592-0, N19858...  
1  [N47996-0, N82719-0, N117066-0, N8491-0, N1237...  
2         [N103852-0, N53474-0, N127836-0, N47925-1]  
3  [N38902-0, N76434-0, N71593-0, N100073-0, N108...  
4  [N76209-0, N48841-0, N67937-0, N62235-0, N6307...

In [81]:
# 뉴스별 사용자에게 선택된 횟수 
def news_readed_count(impression_news):
    candidate_news = collections.defaultdict(int) 

    for news in tqdm(impression_news):
        for cand_news in news:
            # 사용자가 해당 후보 뉴스 읽었으면
            if int(cand_news.split('-')[1]) == 1:
                candidate_news[cand_news.split('-')[0]] += 1

    return dict(candidate_news)

In [84]:
candidate_news = news_readed_count(large_tr_behavior['Impression_News_lst'])



  0%|          | 0/2232748 [00:00<?, ?it/s]

  0%|          | 6421/2232748 [00:00<00:34, 64207.31it/s]

  1%|          | 13858/2232748 [00:00<00:33, 66951.81it/s]

  1%|          | 20976/2232748 [00:00<00:32, 68158.98it/s]

  1%|▏         | 28798/2232748 [00:00<00:31, 70891.51it/s]

  2%|▏         | 36383/2232748 [00:00<00:30, 72309.20it/s]

  2%|▏         | 43948/2232748 [00:00<00:29, 73278.79it/s]

  2%|▏         | 51505/2232748 [00:00<00:29, 73951.40it/s]

  3%|▎         | 59111/2232748 [00:00<00:29, 74571.36it/s]

  3%|▎         | 66216/2232748 [00:00<00:29, 72640.52it/s]

  3%|▎         | 73815/2232748 [00:01<00:29, 73603.26it/s]

  4%|▎         | 81447/2232748 [00:01<00:28, 74395.75it/s]

  4%|▍         | 89007/2232748 [00:01<00:28, 74751.94it/s]

  4%|▍         | 96482/2232748 [00:01<00:28, 74750.63it/s]

  5%|▍         | 104101/2232748 [00:01<00:28, 75172.05it/s]

  5%|▍         | 111581/2232748 [00:01<00:28, 74659.51it/s]

  5%|▌         | 119080/2232748 [00:01<00:28, 74756.

 42%|████▏     | 926979/2232748 [00:12<00:17, 75435.80it/s]

 42%|████▏     | 934524/2232748 [00:12<00:17, 74824.64it/s]

 42%|████▏     | 942072/2232748 [00:12<00:17, 75017.14it/s]

 43%|████▎     | 949576/2232748 [00:12<00:17, 73804.27it/s]

 43%|████▎     | 956962/2232748 [00:12<00:17, 73763.89it/s]

 43%|████▎     | 964543/2232748 [00:12<00:17, 74362.09it/s]

 44%|████▎     | 972310/2232748 [00:12<00:16, 75324.04it/s]

 44%|████▍     | 979951/2232748 [00:13<00:16, 75639.85it/s]

 44%|████▍     | 987520/2232748 [00:13<00:16, 75318.63it/s]

 45%|████▍     | 995056/2232748 [00:13<00:16, 74897.94it/s]

 45%|████▍     | 1002549/2232748 [00:13<00:16, 74635.83it/s]

 45%|████▌     | 1010015/2232748 [00:13<00:16, 73343.40it/s]

 46%|████▌     | 1018014/2232748 [00:13<00:16, 75211.68it/s]

 46%|████▌     | 1025793/2232748 [00:13<00:15, 75965.53it/s]

 46%|████▋     | 1033540/2232748 [00:13<00:15, 76407.76it/s]

 47%|████▋     | 1041191/2232748 [00:13<00:15, 76336.58it/s]

 47%|████▋     | 1

 83%|████████▎ | 1843850/2232748 [00:24<00:05, 74992.49it/s]

 83%|████████▎ | 1851592/2232748 [00:24<00:05, 75704.15it/s]

 83%|████████▎ | 1859212/2232748 [00:24<00:04, 75846.85it/s]

 84%|████████▎ | 1866801/2232748 [00:24<00:04, 75619.38it/s]

 84%|████████▍ | 1874366/2232748 [00:24<00:04, 75590.26it/s]

 84%|████████▍ | 1882032/2232748 [00:24<00:04, 75905.32it/s]

 85%|████████▍ | 1889634/2232748 [00:25<00:04, 75938.35it/s]

 85%|████████▍ | 1897229/2232748 [00:25<00:04, 75715.84it/s]

 85%|████████▌ | 1905173/2232748 [00:25<00:04, 76794.38it/s]

 86%|████████▌ | 1912857/2232748 [00:25<00:04, 76772.85it/s]

 86%|████████▌ | 1920538/2232748 [00:25<00:04, 76269.54it/s]

 86%|████████▋ | 1928202/2232748 [00:25<00:03, 76378.96it/s]

 87%|████████▋ | 1935842/2232748 [00:25<00:03, 76102.46it/s]

 87%|████████▋ | 1943454/2232748 [00:25<00:03, 75656.19it/s]

 87%|████████▋ | 1951042/2232748 [00:25<00:03, 75716.19it/s]

 88%|████████▊ | 1958667/2232748 [00:25<00:03, 75874.86it/s]

 88%|███

In [86]:
# # 뉴스별 사용자에게 선택된 횟수 

# plt.bar(candidate_news.keys(), candidate_news.values())
# plt.yticks(np.arange(min(candidate_news.values()), max(candidate_news.values())+1, 1.0))

# plt.show()

In [92]:
candidate_news.keys() 

dict_keys(['N94157', 'N78699', 'N71090', 'N31174', 'N25587', 'N36266', 'N47925', 'N114935', 'N86258', 'N98178', 'N94572', 'N123077', 'N28902', 'N123968', 'N55816', 'N123683', 'N51048', 'N39403', 'N8701', 'N68624', 'N116694', 'N120192', 'N87838', 'N95254', 'N53859', 'N65783', 'N80272', 'N43186', 'N30899', 'N41547', 'N5098', 'N75494', 'N47643', 'N80444', 'N44037', 'N18947', 'N15184', 'N43755', 'N34269', 'N108379', 'N24701', 'N12453', 'N20990', 'N61065', 'N18539', 'N48015', 'N100073', 'N32154', 'N8599', 'N118589', 'N125974', 'N15420', 'N93674', 'N17053', 'N83421', 'N24150', 'N1328', 'N9262', 'N91151', 'N87684', 'N115443', 'N96729', 'N110627', 'N42649', 'N53474', 'N47257', 'N94850', 'N93955', 'N18509', 'N55944', 'N6293', 'N97935', 'N23086', 'N31323', 'N4371', 'N45522', 'N70923', 'N56873', 'N2750', 'N128057', 'N22895', 'N23954', 'N79233', 'N120312', 'N86931', 'N1923', 'N119988', 'N90968', 'N107088', 'N8719', 'N12483', 'N36828', 'N66780', 'N10627', 'N52617', 'N23748', 'N29441', 'N13279', 'N9

In [107]:
[value for key, value in candidate_news.items() if key == 'N109606']   

[535]

In [111]:
empty_ids = []

for ids in tqdm(no_image_large_tr['0']):
    if [value for key, value in candidate_news.items() if key == ids]  == []:
        empty_ids.append(ids)
        
print(len(empty_ids))
empty_ids



  0%|          | 0/20307 [00:00<?, ?it/s]

  0%|          | 90/20307 [00:00<00:22, 898.73it/s]

  1%|          | 183/20307 [00:00<00:22, 905.29it/s]

  1%|▏         | 282/20307 [00:00<00:21, 927.15it/s]

  2%|▏         | 382/20307 [00:00<00:21, 946.17it/s]

  2%|▏         | 473/20307 [00:00<00:21, 934.91it/s]

  3%|▎         | 568/20307 [00:00<00:21, 939.07it/s]

  3%|▎         | 663/20307 [00:00<00:20, 940.14it/s]

  4%|▎         | 755/20307 [00:00<00:20, 933.52it/s]

  4%|▍         | 850/20307 [00:00<00:20, 936.90it/s]

  5%|▍         | 941/20307 [00:01<00:21, 917.42it/s]

  5%|▌         | 1038/20307 [00:01<00:20, 930.10it/s]

  6%|▌         | 1130/20307 [00:01<00:20, 921.21it/s]

  6%|▌         | 1223/20307 [00:01<00:20, 922.96it/s]

  6%|▋         | 1316/20307 [00:01<00:20, 922.82it/s]

  7%|▋         | 1408/20307 [00:01<00:20, 919.89it/s]

  7%|▋         | 1500/20307 [00:01<00:20, 910.83it/s]

  8%|▊         | 1591/20307 [00:01<00:20, 895.57it/s]

  8%|▊         | 1681/20307 [00

 64%|██████▍   | 13051/20307 [00:13<00:07, 977.74it/s]

 65%|██████▍   | 13149/20307 [00:13<00:07, 975.20it/s]

 65%|██████▌   | 13248/20307 [00:13<00:07, 978.56it/s]

 66%|██████▌   | 13355/20307 [00:13<00:06, 1002.80it/s]

 66%|██████▋   | 13459/20307 [00:14<00:06, 1013.52it/s]

 67%|██████▋   | 13563/20307 [00:14<00:06, 1020.31it/s]

 67%|██████▋   | 13666/20307 [00:14<00:06, 1021.10it/s]

 68%|██████▊   | 13772/20307 [00:14<00:06, 1030.07it/s]

 68%|██████▊   | 13876/20307 [00:14<00:06, 1032.01it/s]

 69%|██████▉   | 13980/20307 [00:14<00:06, 1030.18it/s]

 69%|██████▉   | 14084/20307 [00:14<00:06, 1021.23it/s]

 70%|██████▉   | 14187/20307 [00:14<00:06, 1014.41it/s]

 70%|███████   | 14290/20307 [00:14<00:05, 1016.75it/s]

 71%|███████   | 14393/20307 [00:14<00:05, 1019.50it/s]

 71%|███████▏  | 14495/20307 [00:15<00:05, 1013.34it/s]

 72%|███████▏  | 14598/20307 [00:15<00:05, 1017.02it/s]

 72%|███████▏  | 14700/20307 [00:15<00:05, 1017.78it/s]

 73%|███████▎  | 14803/20307 [00:1

17796


['N42639',
 'N93333',
 'N67829',
 'N41917',
 'N108072',
 'N16951',
 'N3018',
 'N58373',
 'N12731',
 'N26284',
 'N12905',
 'N76727',
 'N1624',
 'N95342',
 'N16588',
 'N127354',
 'N11324',
 'N52135',
 'N78235',
 'N77833',
 'N77682',
 'N91797',
 'N114898',
 'N35553',
 'N94549',
 'N39199',
 'N60246',
 'N4532',
 'N69839',
 'N66010',
 'N117579',
 'N24257',
 'N52760',
 'N98446',
 'N113468',
 'N16414',
 'N5513',
 'N7215',
 'N116092',
 'N82547',
 'N33417',
 'N74412',
 'N114907',
 'N24171',
 'N127081',
 'N80600',
 'N123396',
 'N80894',
 'N72027',
 'N25417',
 'N51071',
 'N47291',
 'N78903',
 'N2687',
 'N97628',
 'N71175',
 'N82611',
 'N71216',
 'N120522',
 'N8204',
 'N11668',
 'N60733',
 'N64293',
 'N63070',
 'N50906',
 'N100713',
 'N13930',
 'N15849',
 'N41182',
 'N84270',
 'N112975',
 'N50912',
 'N112554',
 'N97657',
 'N38868',
 'N119763',
 'N62406',
 'N61547',
 'N70906',
 'N6988',
 'N66309',
 'N45746',
 'N51717',
 'N54452',
 'N59366',
 'N99910',
 'N23015',
 'N127645',
 'N117083',
 'N38958',
 '

empty_ids 17796개(뉴스)는    
behavior(log data)엔 있고, news(크롤링)에 없는 것  
  
즉, 뉴스 정보 (제목, 내용, 날짜, 이미지 등등)은 메타 데이터처럼 있느나,  
유저 log데이터에는 해당 뉴스를 본 적이 없는 것  

* empty_ids  
behavior(log data) impression에서 1  
즉, 사람들이 해당 뉴스를 본 횟수를
(뉴스아이디 : 조회횟수) dict로 저장된 데이터

In [112]:
no_image_large_tr.loc[no_image_large_tr['0'].isin(empty_ids)]

5  \
14      https://assets.msn.com/labs/mind/AAEvtb9.html   
17      https://assets.msn.com/labs/mind/AAJ4lmA.html   
19      https://assets.msn.com/labs/mind/AABDPUi.html   
22      https://assets.msn.com/labs/mind/AAJfTsI.html   
25      https://assets.msn.com/labs/mind/AAISxTL.html   
...                                               ...   
101137  https://assets.msn.com/labs/mind/BBWypeE.html   
101198  https://assets.msn.com/labs/mind/BBWyr4C.html   
101242  https://assets.msn.com/labs/mind/BBWyskJ.html   
101443  https://assets.msn.com/labs/mind/BBWyynd.html   
101483  https://assets.msn.com/labs/mind/BBWz0kE.html   

                                                    Title  \
14      My Corvette Moment? When I Took the Mighty ZR1...   
17      Michigan finally shows some fight, but can’t...   
19      Nashville restaurants: Ms. Cheap rounds up lun...   
22          Top Halloween parties this weekend Oct. 24-27   
25         Swanson Suspended Indefinitely For Broken Hand   
...                                                   ...   
101137  Wrestling: Cyclone Open features ISU-UNI battl...   
101198  'They give us the gift us dedication': Hundred...   
101242  St. Louis County Animal Control revamps volunt...   
101443  Community walk event in Virginia Beach raises ...   
101483      Family asks for help in solving May cold case   

                        Date Image  \
14      2019-07-18T08:00:43Z   NaN   
17      2019-10-20T11:11:58Z   NaN   
19      2019-10-25T16:04:50Z   NaN   
22      2019-10-23T19:11:29Z   NaN   
25      2019-10-16T18:20:09Z   NaN   
...                      ...   ...   
101137  2019-11-10T21:18:12Z   NaN   
101198  2019-11-11T22:48:25Z   NaN   
101242  2019-11-11T04:52:45Z   NaN   
101443  2019-11-10T21:21:16Z   NaN   
101483  2019-11-11T04:38:14Z   NaN   

                                                  Content        0  \
14      It's fitting the C8 Corvette graces the cover ...   N42639   
17      UNIVERSITY PARK, Pa. — Fans on their way out...   N93333   
19      As Ms. Cheap, I'm always looking for lunches t...   N67829   
22      Now that we are in the last weekend before Hal...   N41917   
25      Former UFC strawweight champion, Joanna Jedrze...  N108072   
...                                                   ...      ...   
101137  AMES, Ia. — What makes the Harold Nichols Cy...  N123023   
101198  An estimated 250 people gathered to honor vete...    N2794   
101242  Heather Henderson came to the Olivette center ...  N111343   
101443  A community walk at Mt. Trashmore on Sunday ai...   N17636   
101483  Friends and family of a murder victim is askin...   N82571   

                   1              2  \
14             autos    autossports   
17            sports    more_sports   
19      foodanddrink     newstrends   
22            travel  travelarticle   
25            sports            mma   
...              ...            ...   
101137        sports    more_sports   
101198          news         newsus   
101242          news         newsus   
101443          news   newsbusiness   
101483          news      newscrime   

                                                        3  \
14      My Corvette Moment? When I Took the Mighty ZR1...   
17      Michigan finally shows some fight, but can't s...   
19      Nashville restaurants: Ms. Cheap rounds up lun...   
22          Top Halloween parties this weekend Oct. 24-27   
25         Swanson Suspended Indefinitely For Broken Hand   
...                                                   ...   
101137  Wrestling: Cyclone Open features ISU-UNI battl...   
101198  'They give us the gift us dedication': Hundred...   
101242  St. Louis County Animal Control revamps volunt...   
101443  Community walk event in Virginia Beach raises ...   
101483      Family asks for help in solving May cold case   

                                                        4  \
14      The Big Picture: We take a look at our most me...   
17      UNIV

In [115]:
readed_news = no_image_large_tr.loc[~no_image_large_tr['0'].isin(empty_ids)] 
print(readed_news.shape) 
readed_news.head() 

(2511, 12)


5  \
387    https://assets.msn.com/labs/mind/AAGlKdJ.html   
410    https://assets.msn.com/labs/mind/AAHbS9i.html   
17365  https://assets.msn.com/labs/mind/AAJSXGL.html   
22833  https://assets.msn.com/labs/mind/AAJXjZP.html   
36997  https://assets.msn.com/labs/mind/AAJjvg1.html   

                                                   Title  \
387        2021 Chevy Equinox Caught In The Death Valley   
410    How To Chop An Onion Quickly, Safely + Evenly ...   
17365  Detroit Tigers' most important players for fut...   
22833           Garage A Sinister 1967 Chevy Chevelle SS   
36997  Breaking down how Iowa State and Iowa have rec...   

                       Date Image  \
387    2019-08-26T13:10:29Z   NaN   
410    2019-09-12T10:00:00Z   NaN   
17365  2019-11-13T00:46:59Z   NaN   
22833  2019-11-06T17:18:36Z   NaN   
36997  2019-11-11T15:31:54Z   NaN   

                                                 Content        0  \
387    Still under heavy camo. Two Chevrolet Equinox ...   N24272   
410     Rach is the first to admit that she doesn't c...   N65960   
17365  Because the Detroit Tigers are in the middle ...   N28420   
22833  What a beauty to add to your collection. Of al...  N115298   
36997  It's official: Oskaloosa big man Xavier Foster...    N9136   

                  1                2  \
387           autos        autossuvs   
410    foodanddrink    tipsandtricks   
17365        sports     baseball_mlb   
22833         autos    autosclassics   
36997        sports  basketball_ncaa   

                                                       3  \
387        2021 Chevy Equinox Caught In The Death Valley   
410    How To Chop An Onion Quickly, Safely + Evenly ...   
17365  Detroit Tigers' most important players for fut...   
22833           Garage A Sinister 1967 Chevy Chevelle SS   
36997  Breaking down how Iowa State and Iowa have rec...   

                                                       4  \
387                              Still under heavy camo.   
410    Forget the "proper" way to chop an onion   thi...   
17365  Detroit Tigers pitcher Michael Fulmer has beco...   
22833           What a beauty to add to your collection.   
36997  With less than three weeks left until Foster a...   

                                                       6  \
387    [{"Label": "Chevrolet Equinox", "Type": "V", "...   
410                                                   []   
17365  [{"Label": "Michael Fulmer", "Type": "P", "Wik...   
22833  [{"Label": "Chevrolet Chevelle", "Type": "V", ...   
36997                                                 []   

                                                       7  
387                                                   []  
410                                                   []  
17365  [{"Label": "Michael Fulmer", "Type": "P", "Wik...  
22833                                                 []  
36997                                                 []

In [123]:
readed_news['img'] = readed_news.apply(lambda x: 0 if x.isnull else 1 )
readed_news.head()

/Users/baeyuna/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


5  \
387    https://assets.msn.com/labs/mind/AAGlKdJ.html   
410    https://assets.msn.com/labs/mind/AAHbS9i.html   
17365  https://assets.msn.com/labs/mind/AAJSXGL.html   
22833  https://assets.msn.com/labs/mind/AAJXjZP.html   
36997  https://assets.msn.com/labs/mind/AAJjvg1.html   

                                                   Title  \
387        2021 Chevy Equinox Caught In The Death Valley   
410    How To Chop An Onion Quickly, Safely + Evenly ...   
17365  Detroit Tigers' most important players for fut...   
22833           Garage A Sinister 1967 Chevy Chevelle SS   
36997  Breaking down how Iowa State and Iowa have rec...   

                       Date Image  \
387    2019-08-26T13:10:29Z   NaN   
410    2019-09-12T10:00:00Z   NaN   
17365  2019-11-13T00:46:59Z   NaN   
22833  2019-11-06T17:18:36Z   NaN   
36997  2019-11-11T15:31:54Z   NaN   

                                                 Content        0  \
387    Still under heavy camo. Two Chevrolet Equinox ...   N24272   
410     Rach is the first to admit that she doesn't c...   N65960   
17365  Because the Detroit Tigers are in the middle ...   N28420   
22833  What a beauty to add to your collection. Of al...  N115298   
36997  It's official: Oskaloosa big man Xavier Foster...    N9136   

                  1                2  \
387           autos        autossuvs   
410    foodanddrink    tipsandtricks   
17365        sports     baseball_mlb   
22833         autos    autosclassics   
36997        sports  basketball_ncaa   

                                                       3  \
387        2021 Chevy Equinox Caught In The Death Valley   
410    How To Chop An Onion Quickly, Safely + Evenly ...   
17365  Detroit Tigers' most important players for fut...   
22833           Garage A Sinister 1967 Chevy Chevelle SS   
36997  Breaking down how Iowa State and Iowa have rec...   

                                                       4  \
387                              Still under heavy camo.   
410    Forget the "proper" way to chop an onion   thi...   
17365  Detroit Tigers pitcher Michael Fulmer has beco...   
22833           What a beauty to add to your collection.   
36997  With less than three weeks left until Foster a...   

                                                       6  \
387    [{"Label": "Chevrolet Equinox", "Type": "V", "...   
410                                                   []   
17365  [{"Label": "Michael Fulmer", "Type": "P", "Wik...   
22833  [{"Label": "Chevrolet Chevelle", "Type": "V", ...   
36997                                                 []   

                                                       7  img  
387                                                   []  NaN  
410                                                   []  NaN  
17365  [{"Label": "Michael Fulmer", "Type": "P", "Wik...  NaN  
22833                                                 []  NaN  
36997                                                 []  NaN

In [104]:
no_image_large_tr.tail()  

5  \
101490  https://assets.msn.com/labs/mind/BBWz2YQ.html   
101512  https://assets.msn.com/labs/mind/BBWzBjj.html   
101513  https://assets.msn.com/labs/mind/BBWzBrF.html   
101523  https://assets.msn.com/labs/mind/BBWzQYV.html   
101524  https://assets.msn.com/labs/mind/BBWzQnK.html   

                                                    Title  \
101490  Four siblings, all veterans, celebrate Veteran...   
101512  New buzz on possibility of Major League Soccer...   
101513  Still teaching at 95, Jimmy Carter draws devot...   
101523  Family says 13-year-old Broadway star died fro...   
101524  St. Dominic soccer player tries to kick cancer...   

                        Date Image  \
101490  2019-11-11T13:30:00Z   NaN   
101512  2019-11-11T05:44:15Z   NaN   
101513  2019-11-11T06:16:27Z   NaN   
101523  2019-11-11T10:45:42Z   NaN   
101524  2019-11-11T05:11:23Z   NaN   

                                                  Content        0          1  \
101490  Those who pay a visit to the solemn World War ...   N62913  lifestyle   
101512  Charlotte appears to be in the driver’s seat...   N98362     sports   
101513  PLAINS, Ga. (AP) — The pilgrims arrive early...  N109606       news   
101523  Eyewitness News is learning what caused the de...   N61455  lifestyle   
101524  Sometimes, what happens on the sidelines can b...  N115249     sports   

                      2                                                  3  \
101490  lifestylefamily  Four siblings, all veterans, celebrate Veteran...   
101512       soccer_mls  New buzz on possibility of Major League Soccer...   
101513           newsus  Still teaching at 95, Jimmy Carter draws devot...   
101523  lifestylefamily  Family says 13-year-old Broadway star died fro...   
101524      more_sports  St. Dominic soccer player tries to kick cancer...   

                                                        4  \
101490  Those who pay a visit to the solemn World War ...   
101512  Charlotte appears to be in the driver's seat f...   
101513  PLAINS, Ga. (AP)   The pilgrims arrive early a...   
101523                                                NaN   
101524  Sometimes, what happens on the sidelines can b...   

                                                        6  \
101490  [{"Label": "Veterans Day", "Type": "H", "Wikid...   
101512  [{"Label": "Major League Soccer", "Type": "O",...   
101513  [{"Label": "Jimmy Carter", "Type": "P", "Wikid...   
101523  [{"Label": "Broadway theatre", "Type": "F", "W...   
101524                                                 []   

                                                        7  
101490                                                 []  
101512  [{"Label": "Major League Soccer", "Type": "O",...  
101513  [{"Label": "Jimmy Carter", "Type": "P", "Wikid...  
101523                                                 []  
101524                                                 []